blog/2024-07-3-uppy-4.0.md

```diff
@@ -0,0 +1,408 @@
+  you initiate a multipart upload, there is no expiry; you must explicitly
+  complete or stop the multipart upload.
+- Begin an upload before you know the final object size – You can upload an
+  object as you are creating it.
+
+However, the downside is request overhead, as it needs to do creation, signing,
+and completion requests besides the upload requests. For example, if you are
+uploading files that are only a couple kilobytes with a 100ms roundtrip latency,
+you are spending 400ms on overhead and only a few milliseconds on uploading.
+This really adds up if you upload a lot of small files.
+
+AWS, and generally the internet from what we found, tend to agree that **you
+don’t want to use multipart uploads for files under 100MB**. But this sometimes
+puts users of our libraries in an awkward position, as their end users may not
+only upload very large files, or only small files. In this case a portion of
+their users get a subpar experience.
+
+---
+
+We’ve merged the two plugins into `@uppy/aws-s3` with a new
+[`shouldUseMultipart`](/docs/aws-s3/#shouldusemultipartfile) option! By default
+it switches to multipart uploads if the file is larger than 100MB. You can pass
+a `boolean` or a function to determine it per file.
+
+## React hooks
+
+People working with React are more likely to create their own user interface on
+top of Uppy than those working with “vanilla” setups. Working with our pre-build
+UI components is a plug-and-play experience, but building on top of Uppy’s state
+with React primitives has been tedious.
+
+To address this we’re introducing to new hooks: `useUppyState` and
+`useUppyEvent`. Thanks to the TypeScript rewrite, we can now do powerful
+inference in hooks as well.
+
+### `useUppyState(uppy, selector)`
+
+Use this hook when you need to access Uppy’s state reactively.
+
+```ts
+import { useState } from 'react';
+import Uppy from '@uppy/core';
+import { useUppyState } from '@uppy/react';
+
+// IMPORTANT: passing an initializer function
+// to prevent Uppy from being recreated on every render.
+const [uppy] = useState(() => new Uppy());
+
+const files = useUppyState(uppy, (state) => state.files);
+const totalProgress = useUppyState(uppy, (state) => state.totalProgress);
+// We can also get specific plugin state.
+// Note that the value on `plugins` depends on the `id` of the plugin.
+const metaFields = useUppyState(
+	uppy,
+	(state) => state.plugins?.Dashboard?.metaFields,
+);
+```
+
+You can see all the values you can access on the
+[`State`](https://github.com/transloadit/uppy/blob/dab8082a4e67c3e7f109eacfbd6c3185f117dc60/packages/%40uppy/core/src/Uppy.ts#L156)
+type. If you are accessing plugin state, you would have to look at the types of
+the plugin.
+
+```ts
+import type { State } from '@uppy/core';
+```
+
+### `useUppyEvent(uppy, event, callback)`
+
+Listen to Uppy [events](/docs/uppy/#events) in a React component.
+
+Returns an array of which the first item is an array of results from the event.
```

Comment:
Not sure how to re-phrase that, but I find the current wording hard to parse.

- maybe instead of array, we could say pair?
- "the first item is an array of results from the event" – but only if an event has been emitted